src/cmap/conn/mod.rs

```diff
@@ -70,6 +71,8 @@ pub(crate) struct Connection {
 
     #[derivative(Debug = "ignore")]
     handler: Option<Arc<dyn CmapEventHandler>>,
+
+    server_api: Option<ServerApi>,
```

Comment:
Out of curiosity, is putting the server API version on a connection a requirement of the spec? It's not immediately obvious to me why this would be necessary, so I'd like to understand better why this is being done

Reply:
Authentication commands and the handshake bypass the operation executor and create their own command objects. This means that the "lowest" common place to inject the server API version would be `Connection::send_command`, requiring us to store the `ServerApi` object on the connection.

Reply:
Right; I'm wondering if it would make more sense to just having the calling context set the API version on a command before calling `send_command` like we do with some other properties. This seems like more of a Rust driver-specific architecture question, so I'm going to tag in @patrickfreed and @isabelatkinson to see if they have any opinions on the options here.

Reply:
From what I can see, moving this from `send_command` to the calling context would require us to pass server api info into the auth mechanism implementations and append it to the commands there.

On the other hand, if (and this is a big if at the moment) we were to allow users to override the declared API version on a more granular level (e.g. on a per-operation basis), automatically setting the api version on the connection might not be the right call, but as is I think it's the most logical place to keep this. I'll defer to your judgment on this since, but I might require some help if you decide that this shouldn't be in `send_command`.

Reply:
I think Patrick's suggestion is probably the most consistent with how we're using commands and connections in the driver. The way things work right now, whatever is passed to `send_command` is exactly what's sent to the server; I'm not sure I want to go down the route where the connection modifies commands before sending them to the server, as then it gets trickier to reason about what's sent to the server from outside of the `send_command` function.

Reply:
So to support this approach, we'd need to do the store the `ServerApi` value on the `ConnectionEstablisher` and pass it through through to `handshaker.handshake` and `credential.authenticate_stream` as necessary (https://github.com/mongodb/mongo-rust-driver/blob/master/src/cmap/establish/mod.rs#L12). I think that in conjunction with the existing storage of `ServerApi` on `ClientInner` and setting in `execute_operation`, this should cover the cases we need without having to store anything on the `Connection` object itself.

Reply:
I've made the changes in a3c6a9b347452bb97b35ecf4305b5d4fa9b28bc3. One notable difference is that I didn't pass `server_api` through `handshaker.handshake`. Since the `isMaster` command is generated when the Handshaker instance is created, we can also apply server_api parameters there and don't have to worry about them later. I think that takes care of everything, but please let me know if I missed anything.